rfc/text/wamp-terminology.md

```diff
@@ -0,0 +1,76 @@
+This is to support the overarching goal of WAMP to free application developers from restrictions when
+building distributed applications, both at the network level, and when choosing or switching the WAMP client
+libraries or routers used.
+
+Our goal is to maximize **User Choice and Experience** for developers of WAMP based applications,
+both formally (open protocol and open source) as well as practically (switching costs).
+
+
+### Fundamental
+
+{align="left"}
+| Term                              | Definition                                                                                              |
+|-----------------------------------|---------------------------------------------------------------------------------------------------------|
+| *User*                            | A user runs *Client*s or *Router*s which implement WAMP                                                 |
+| *Client*                          | A program run by a *User* with application code using WAMP for application level communication          |
+| *Router*                          | A program run by a *User* with middleware code using WAMP to provide application routing services       |
+| *Peer*                            | A WAMP *Client* or *Router*. An implementation might embed, provide or use both roles                   |
+| *Realm*                           | Isolated WAMP URI namespace, routing and administrative domain, optionally protected by **AA**          |
+| *Transport*                       | A message-based, reliable, ordered, bidirectional (full-duplex) channel over which *Peers* communicate  |
+| *Connection*                      | An underlying entity (if any) carrying the *Transport*, e.g. a network connection, pipe, queue or such  |
+| *Session*                         | Transient conversation between a *Client* and a *Router* on a *Realm* over a *Transport*                |
+| *Message*                         | Indivisible unit of information transmitted between peers                                               |
+| *Serializer*                      | A *Serializer* encodes WAMP messages with application payloads into bytes strings for transport         |
+
+
+### Authentication and Authorization (AA)
+
+{align="left"}
+| Term                               | Definition                                                                                                                     |
+|------------------------------------|--------------------------------------------------------------------------------------------------------------------------------|
+| *Authentication*                   | Establishes the identity of a *Session* on a *Realm*                                                                           |
+| *Principal*                        | Once authenticated, *Session*s identify under a *Principal*                                                                    |
```

Comment:
how about: "A user that can be authenticated on a realm, under a role"

https://en.wikipedia.org/wiki/Principal_(computer_security)

Reply:
technically, the principal is the `authid`  - and that is defined/relative to the realm.

Reply:
I think it's a has-a relationship between Principal and `authid`: A Principal **has** an authid.

Your suggestion of "A user that can be authenticated on a realm, under a role" seems fine, exception that I would substitute "user" with "entity", because *User* was previously defined as something else.

Reply:
mmh .. not sure ... my view:

> I think it's a has-a relationship between Principal and authid: A Principal has an authid.

No, the authid is just a string identifier for the principal - it is 1:1

The user-principal relation is 1:n, since one user can have different principals on realms

"User" is defined as "A person (or organization) running a WAMP *Client* or *Router*"

"Entity": is undefined and arbitrary

Reply:
> No, the authid is just a string identifier for the principal - it is 1:1

What I meant to say is that `authid` is an attribute of a Principal (technically, a has-a relationship with a String). That's not the same as saying the principal is the authid.

This class diagram models the way I understand it:
```mermaid
classDiagram
    direction LR
    User "1" -- "*" Principal
    Principal "*" o-- "1" Role
    class User {
        <<actor>>
    }
    class Principal {
        +authId: String
        +authRole: String
    }
```

Some apps will associate multiple roles to the same Principal, but WAMP is currently limited to a single role. I'm aware this has been brought up already in an issue.

From [Microsoft](https://learn.microsoft.com/en-us/windows-server/identity/ad-ds/manage/understand-security-principals):
> A security principal is any entity that can be authenticated by the operating system, such as a user account, a computer account, or a thread or process that runs in the security context of a user or computer account, or the security groups for these accounts.

In the case of WAMP, if we apply this definition, a security principal can be a user account or a role (security group).

Using Microsoft's definition, the class diagram should really be:
```mermaid
classDiagram
    direction LR
    Principal <|-- Account
    Principal <|-- Role
    User "1" -- "*" Account
    Account "*" o-- "1" Role
    class User {
        <<actor>>
    }
    class Account {
        +authId: String
        +authRole: String
    }
```

Reply:
A WAMP Role can't be authenticated, so perhaps it should not be considered a Principal according to MS's definition. I don't know how one "authenticates" a security group. It's usually the account that's authenticated, and an account is assigned to one or more security groups. One could **authorize** a security group to perform an action, though.

Reply:
No, I disagree: a Principal for a User only is defined in a Realm - it doesn't exist independently from a Realm

Reply:
> What I meant to say is that authid is an attribute of a Principal (technically, a has-a relationship with a String). That's not the same as saying the principal is the authid.

Ok, that makes sense. "has a". But the relation is only defined on a Realm. So:

User 1 --- Realm --- 1|0 Principal (authid)

A User can exist without a Principal (== authid) on a Realm.

Reply:
> A WAMP Role can't be authenticated, so perhaps it should not be considered a Principal 

Who said a Role would be considered a Principal?

Reply:
> One could authorize a security group to perform an action, though.

well, that's exactly how it is defined: the Role on a Realm, the URI and the Action determine the Access Controls

Reply:
> Who said a Role would be considered a Principal?

The Microsoft definition considers "security groups" to be a kind of Security Principal, and I think a Role could be considered a security group.

If you don't agree with MS's inclusion of "security groups" a kind of Security Principal, that's fine, as long as we define "Principal" in clear manner and be consistent in its usage.

Reply:
How about this?
Principal: A user account associated with a *Role*, having a string identifier (`authid`), that can be authenticated under a *Realm*

Reply:
I think the MS concept is mixing two concepts: Principals and Roles (what they call Groups). If a Principal has always one Role, and a Role is a Principal, this is mixing 2 different kinds of concepts. I am -1 on this.

----

> as long as we define "Principal" in clear manner and be consistent in its usage

following roughly the MS definition:

**Principal**: A security principal (`authid`) is any *User* that can be authenticated on a *Realm* (`realm`) that runs in the security context of a *Role* (`authrole`) on the *Realm*.

Reply:
Alright, I'll meet you halfway and agree to your latest definition with these grammatical changes:

Principal: A security principal (authid) is any User that can be authenticated under a Realm (realm) that runs in the security context of a Role (authrole) within that Realm.

"authenticate **under** a Realm" sounds better to my ear in the same way as "file that **under** receipts".

Perhaps the preposition "on" makes sense in German, but its doesn't sound right in English for this particular instance (neither in French, for that matter).

Reply:
great! let's go with your last version.

fwiw, yeah, that wrong preposition is likely due to my native german tongue;) here is how it would read:

"Prinzipal: Ein Sicherheitsprinzipal ist jeder Nutzer der für einen Verwaltungsbereich authentifiziert ist und welcher unter dem Sicherheitskontext einer Rolle in diesem Verwaltungsbereich agiert."

this is definitely not entry level language. but anyways, it made be recognize 1 tiny more grammar glitch:

> Principal: A security principal (authid) is any User that can be authenticated under a Realm (realm) that runs in the security context of a Role (authrole) within that Realm.

the 2nd "that" could be read as to relate to Realm - where it should relate to User. ?? Maybe we should replace 2nd "that" with "and"? (I did that in the german version btw)

anyways, call me crazy, but thank you very much for caring about these details! I do as well. in particular in a terminology chapter.

Reply:
updated: https://github.com/wamp-proto/wamp-proto/pull/454/commits/157eb8c90a09bda54209eb0fb793470025246b92#diff-270c0c150b78b0712506f081bd98b0e462353bcaff0cdf77e2061b9cbe244989R39

Reply:
I see what you mean about the possible confusion, and I like your version better in your update.